curp/src/server.rs

```diff
@@ -627,87 +669,77 @@ impl<C: 'static + Command, CE: 'static + CommandExecutor<C>> Protocol<C, CE> {
         let cmd = p.cmd().map_err(|e| {
             tonic::Status::invalid_argument(format!("propose cmd decode failed: {}", e))
         })?;
-        self.spec
-            .map_lock(|mut spec| {
-                let is_conflict = Self::is_spec_conflict(&spec, &cmd);
-                spec.push_back(cmd.clone());
-                debug!("insert cmd {:?} to spec pool", cmd.id());
-
-                match (is_conflict, is_leader) {
-                    // conflict and is leader
-                    (true, true) => async {
-                        // the leader needs to sync cmd
-                        match self.sync_to_others(self_term, &cmd, true) {
-                            Ok(notifier) => {
-                                notifier.notify(1);
-                                ProposeResponse::new_error(
-                                    is_leader,
-                                    self_term,
-                                    &ProposeError::KeyConflict,
-                                )
-                            }
-                            Err(_err) => ProposeResponse::new_error(
-                                is_leader,
-                                self.state.read().term,
-                                &ProposeError::SyncedError("Sync cmd channel closed".to_owned()),
-                            ),
-                        }
-                    }
-                    .left_future()
-                    .left_future(),
-                    // conflict but not leader
-                    (true, false) => async {
+
+        let is_conflict = {
+            let mut spec = self.spec.lock();
+            let is_conflict = spec.has_conflict(&cmd);
+            // push the new command to the spec pool
+            if !spec.push(cmd.clone()) {
+                // If the command has already been committed and executed, return OK. This will happen when the command has already been synced and executed before the propose has arrived to the follower.
+                return Ok(tonic::Response::new(
+                    ProposeResponse::new_empty(false, self.state.read().term).map_err(|e| {
+                        tonic::Status::internal(format!("encode or decode error, {}", e))
+                    })?,
+                ));
+            }
+            is_conflict
+        };
+
+        let resp = match (is_conflict, is_leader) {
+            // conflict and is leader
+            (true, true) => {
+                // the leader needs to sync cmd
+                match self.sync_to_others(self_term, &cmd, true) {
```

Comment:
We can't change the logic this way, because we must make sure the first-coming request sync to others first. The order is kept for by holding the spec lock.